// inversion-of-control/lib/decorators/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Service = Service;
exports.Injectable = Injectable;
exports.Controller = Controller;
exports.Repository = Repository;
exports.isService = isService;
exports.isController = isController;
exports.isRepository = isRepository;
require("reflect-metadata");
const SERVICE_METADATA_KEY = Symbol('SERVICE_METADATA_KEY');
const CONTROLLER_METADATA_KEY = Symbol('CONTROLLER_METADATA_KEY');
const REPOSITORY_METADATA_KEY = Symbol('REPOSITORY_METADATA_KEY');
function Service() {
    return function (target) {
        Reflect.defineMetadata(SERVICE_METADATA_KEY, true, target);
    };
}
function Injectable() {
    return function (target) {
        Reflect.defineMetadata('design:paramtypes', Reflect.getMetadata('design:paramtypes', target) || [], target);
    };
}
function Controller() {
    return function (target) {
        Reflect.defineMetadata(CONTROLLER_METADATA_KEY, true, target);
    };
}
function Repository() {
    return function (target) {
        Reflect.defineMetadata(REPOSITORY_METADATA_KEY, true, target);
    };
}
function isService(target) {
    return Reflect.getMetadata(SERVICE_METADATA_KEY, target) === true;
}
function isController(target) {
    return Reflect.getMetadata(CONTROLLER_METADATA_KEY, target) === true;
}
function isRepository(target) {
    return Reflect.getMetadata(REPOSITORY_METADATA_KEY, target) === true;
}
